### ISSUE TITLE
Identify nodes by UUID for coordination and messaging instead of IPs

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#6403]
Scylla allows replacing a dead node with a node which has the same IP as the dead node.

During replace operation we don't want messages targeted to the old node to arrive to the new node. This could violate consistency invariants, as the new node may not yet be fully synchronized. Refs https://github.com/scylladb/scylla/issues/4648

At the same time, we do want some messages to arrive to the new node. For example, data synchronization algorithm requires that current requests arrive to the new node while it's being synchronized.

I propose that we identify nodes by host IDs in the ring instead of IPs. Messaging layer will allow sending to a host ID instead of to an IP (it will still allow to send by IP). Messaging layer will keep the translation. When a connection is established between nodes, messaging layer learns about node's host ID during early negotiation before any message can be sent through that connection. This will make sure that we never send any message to the wrong node.

This will also remove the need to have special-casing in the coordinator to disambiguate between IPs. This means we won't need `abstract_replication_strategy::get_natural_endpoints_without_node_being_replaced()`.

**Comment by dorlaor** [comment]
It's a good idea. Can you pls dig into C*'s JIRA to see if they
have such an enhancement request? Worth to contribute
such a request if one doesn't exist.

If this has effects on our configuration file, worth to plan it and
also have a compatibility/transition mode with IPs

**Comment by kostja** [comment]
Do you suggest UUIDs are specified by the user or created transparently to the user by Scylla? In the latter case, how do you suggest obsolete UUIDs are expired?

**Comment by tgrabiec** [comment]
@kostja Transparently. We already assign such IDs (see `system_keyspace::get_local_host_id()`). 

By expiring, do you mean garbage collecting unused ID->IP mappings? We only need to track IDs which are present in the ring (token_metadata), only those will be contacted by coordinators. We already persist this mapping in system tables. When a node is removed from the ring, the mapping goes away.

There'd be also an in-memory mapping kept by messaging service, this goes away when a node disconnects.

**Comment by dorlaor** [comment]
https://issues.apache.org/jira/browse/CASSANDRA-15823

**Comment by kbr-** [comment]
I support this cause.

We could take this even further: decouple the messaging layer completely from other parts of the system, make it pluggable, like `raft::server` has a pluggable `rpc` interface.

This could allow us to do fun stuff like starting multiple Scylla nodes that communicate through a fake network... in a single process. Would make it possible to run cluster tests inside the unit testing framework!

**Comment by kostja** [comment]
We should be able to do it as the first application of Raft based configuration changes, so I'd put this ticket into that train.

**Comment by avikivity** [comment]
> I support this cause.


I support your support.


> 
> We could take this even further: decouple the messaging layer completely from other parts of the system, make it pluggable, like `raft::server` has a pluggable `rpc` interface.
> 
> This could allow us to do fun stuff like starting multiple Scylla nodes that communicate through a fake network... in a single process. Would make it possible to run cluster tests inside the unit testing framework!

We are ever closer to this, thanks to @xemul efforts against global variables.

(but this can be done even with IP addresses, by having a mapping between fake IP addresses and in-process nodes)

**Comment by tgrabiec** [comment]
Why do we have to employ raft?

Unique host IDs can be assigned without coordination because it's the ID
owner who assigns it to himself. We already do it.

**Comment by kostja** [comment]
I imagined we would simply want to run all updates of a system table through raft, regardless of whether or not they can be run without it. E.g. we could store short server ids in addition to long uuids for use in messaging and WAL.

**Comment by slivne** [comment]
@tgrabiec / @bhalevy please break this first to a list of items / modules that need to be u[dated to make this done - we can then work on assigning the parts to different owners

**Comment by tgrabiec** [comment]
Here's a rough implementation plan.

1. Introduce a new type which represents the concept of a `host`. A host moves in a linear fashion over a well defined set of states (bootstrapping -> normal -> leaving -> left). It can be assumed to contain a monotonic set of writes (doesn't lose data) for the data ranges it owns. Unlike `inet_address`, which can refer to different hosts at different points in time. 

```
using host_id = utils::UUID;

class host {
   host_id id;
   inet_address addr;
};
```

2. Initialize `messaging_service` instance with a local host id attached to it. Make sure host id is generated/loaded before messaging service is initialized. See `get_local_host_id()`.

3. Advertise host id through messaging_service when RPC connection is established so that both sides know their host ids before any RPC is sent. 

2. messaging_service will contain a structure called "host routing", which contains a mapping between host_id and inet_address.

2. Storage service populates the host routing from system tables on boot (see `load_host_ids()`). Routing is also added (in memory, transient) when a peer connects to the messaging service server.

4. Extend `netw::msg_addr` to allow sending to a `host_id` (when set). When there is no routing for a given host_id, the call fails. `messaging_service::get_source()` creates a msg_addr with host_id populated from client_info.

5. Change the node replacement procedure to not reuse the host id when bootstrapping the replacing node.

5. Replace `inet_address` uses in the gossiper with `host_id`. Host id is obtained from `msg_addr` by the server. Existing handlers still need `inet_address` (mainly to persist routing), so handlers should be changed to work in terms of the `host` object, which has both host_id and `inet_address`.

5.  Replace uses of `inet_address` in `token_metadata` with `host`. We can drop the uses of `token_metadata::update_host_id()`.

5. Replace `inet_address` uses in storage_proxy with `host_id`.

5. Drop the workarounds in `get_natural_endpoints_without_node_being_replaced()`.

**Comment by bhalevy** [comment]
@tgrabiec the above sounds good.
@asias please also review the above plan.

**Comment by avikivity** [comment]
I think most of references to hosts should be with a host*, not a host_id. This avoids lookups (but on the other hand, requires care not to delete a host before all references to it are gone).

**Comment by tgrabiec** [comment]
@avikivity We can always have  

```
using host_ptr = lw_shared_ptr<host>;
```

Which lookups you want to avoid? In the messaging_service's routing table?

**Comment by avikivity** [comment]
> @avikivity We can always have
> 
> ```
> using host_ptr = lw_shared_ptr<host>;
> ```

That triggers huge amounts of lw_shared_ptr overhead and destructor calls, killing many optimizations in the small containers we use.

> 
> Which lookups you want to avoid? In the messaging_service's routing table?

Yes, I assume (but didn't check) there are also lookups to see if the node is up or down.

**Comment by tgrabiec** [comment]
> > @avikivity We can always have
> > ```
> > using host_ptr = lw_shared_ptr<host>;
> > ```
> 
> That triggers huge amounts of lw_shared_ptr overhead and destructor calls, killing many optimizations in the small containers we use.

Maybe it's worth to sacrifice some performance for safety, if it's negligible. What kind of optimizations are killed?

> 
> > Which lookups you want to avoid? In the messaging_service's routing table?
> 
> Yes, I assume (but didn't check) there are also lookups to see if the node is up or down.

**Comment by slivne** [comment]
@avikivity comments on https://github.com/scylladb/scylla/issues/6403#issuecomment-771564667 - we need to move forward with this as we are hitting other issues

**Comment by eliransin** [comment]
@kostja reading the above conversation, I still don't understand what raft has to do with it.
Can you elaborate a little bit about that?

**Comment by kostja** [comment]
If nodes use raft RPC to communicate, this issue is a given. Nothing special needs to be done about it.

**Comment by kostja** [comment]
Today the nodes have not one, but *two* globally unique host identifiers: host id and raft group 0 id. It is encouraged that all communication that needs to happen for entire cluster is performed by writing to system tables managed by Raft and reading the information from these system tables using Raft read barrier. This includes topology, snitch, schema, and any other information concerning the entire cluster. If communication is done via writing into Raft Group 0 state machine, it not just accounts for this issue, but durability, linearizability, correctness in presence of topology changes, etc. As such I don't think anything needs to be done with this issue, except that every susbsystem needs to stop using RPC to exchange cluster state.

On differences between Raft Group 0 id and host id please track this ticket: https://github.com/scylladb/scylla/issues/9839

**Comment by gleb-cloudius** [comment]
Access to many (all?) of those things you propose to read using raft is on the fast path and thus it is completely out of the question to use read_barrier() that have high latency to read them.

**Comment by tgrabiec** [comment]
@kostja Even if we get the `host id <-> IP` mapping from raft somehow, this issue is also about using host id in the RPC layer so that the target of RPC call drops it if it has a different host id (but same IP). This RPC could be, for example, a mutate verb. We also have to change the coordinator (storage_proxy) code to use host ids instead of IPs.

**Comment by mykaul** [comment]
@kostja - which teams own this feature and can scope it and estimate the effort?

**Comment by kostja** [comment]
@mykaul There is no clearly defined scope in this ticket. On one hand, all Raft messages already identify peers with Raft IDs. So for all future work which uses Raft messaging this ticket is done. On the other, we have a lot of legacy RPC that serves reads and writes to eventually  consistent tables, and legacy cluster management/repair RPC. It is still uses the old API. Fixing it is a lot of work. So for it, the ticket needs to be split into smaller ones, and I guess we should begin with using host ids instead of IPs to address endpoints. Generally I'd love to own this scope but a) there is no capacity in the team at the moment b) there is no clearly defined scope between teams today in general. A lot of related work has been done by @asias and @eliransin  team.

**Comment by kbr-** [comment]
>  propose that we identify nodes by host IDs

BTW. Host IDs are transferred during the replace operation, so I don't think they are good enough. We need a unique immutable node ID. The Raft server ID is an example of such.

**Comment by tgrabiec** [comment]
On Wed, Nov 2, 2022 at 1:55 PM Kamil Braun ***@***.***> wrote:

> propose that we identify nodes by host IDs
>
> BTW. Host IDs are transferred during the replace operation, so I don't
> think they are good enough. We need a unique immutable node ID. The Raft
> server ID is an example of such.
>
Yes. Part of the original proposal was to ensure that replace doesn't reuse
host IDs. We can introduce another kind of ID, which is truly unique. I
think we will regret not making host ID truly unique now, it's a legacy
which will add to even more confusion.

**Comment by avikivity** [comment]
Can we change replace not to transfer host ID?

**Comment by avikivity** [comment]
ps the raft group 0 server ID should have reused the host ID.

**Comment by kostja** [comment]
it didn't exactly because of the scenario with replace. 
https://github.com/scylladb/scylladb/issues/10792
https://github.com/scylladb/scylladb/issues/9839

**Comment by avikivity** [comment]
Well, maybe we should fix replace.

**Comment by mykaul** [comment]
> #12250 just provides the prerequisite of unique host_id. There are a lot of places where we currently use the nodes ip address that we should convert to host_id. We should probably enumerate them in separate issues so we can address them one at a time. We definitely can't target all of that for 5.2.

ACK - moved to 5.3.

**Comment by kbr-scylla** [comment]
Whatever was needed to unblock Raft-topology was done. (token_metadata, effective_replication_map)
The rest is something we should solve eventually, but not a blocker.
Moving to post-release backlog
